README.md
## `omit-empty-es`

This is a (_temporary_) **fork** of the [`omit-empty`](https://github.com/jonschlinkert/omit-empty) library by [Jon Schlinkert](https://github.com/jonschlinkert) and is meant to produce a bundle that is properly consumed by modern build tools (Babel, Webpack, etc.).

> Hopefully we can merge those changes upstream and this repository/package won't be necessary anymore.

### Install

```bash
npm install --save omit-empty-es
```

```bash
yarn add omit-empty-es
```

### Usage

```js
import omitEmpty from 'omit-empty-es';
```


CHANGELOG.md
## 1.1.0

## 1.2.0

### Minor Changes

- [#38](https://github.com/emmenko/omit-empty-es/pull/38) [`c427f1b`](https://github.com/emmenko/omit-empty-es/commit/c427f1b748fbfcf738160abbd42d046932ce72e3) Thanks [@emmenko](https://github.com/emmenko)! - Update dependencies and repository tooling

### Patch Changes

- [#49](https://github.com/emmenko/omit-empty-es/pull/49) [`d45b316`](https://github.com/emmenko/omit-empty-es/commit/d45b316dfa656deab4d7da357aad6b5caa110a44) Thanks [@emmenko](https://github.com/emmenko)! - Update dependencies

## 1.1.3

### Patch Changes

- [`e3e0c6a`](https://github.com/emmenko/omit-empty-es/commit/e3e0c6af13ab7defb57a717df965b9ff0271a628) [#10](https://github.com/emmenko/omit-empty-es/pull/10) Thanks [@renovate](https://github.com/apps/renovate)! - chore(deps): update all dependencies

* [`a1caace`](https://github.com/emmenko/omit-empty-es/commit/a1caace6bb920b1ecaf295548356e29d81dabd06) Thanks [@emmenko](https://github.com/emmenko)! - Use loosen range for babel deps

## 1.1.2

### Patch Changes

- [`e0e7cc1`](https://github.com/emmenko/omit-empty-es/commit/e0e7cc18a3d8d3cc0c3ee25595c01cfbbdbce8b3) Thanks [@emmenko](https://github.com/emmenko)! - Remove postinstall script

## 1.1.1

### Patch Changes

- [`4abfbae`](https://github.com/emmenko/omit-empty-es/commit/4abfbaea37c6d5daa1fbb11873c5d999243649e5) Thanks [@emmenko](https://github.com/emmenko)! - Fix wrong production bundles

### Minor Changes

- [`f968e80`](https://github.com/emmenko/omit-empty-es/commit/f968e80fa3d15ce9be9b41dbee4457d988637c48) [#1](https://github.com/emmenko/omit-empty-es/pull/1) Thanks [@emmenko](https://github.com/emmenko)! - Use TypeScript, improve CI tools.

## 1.0.3

- bundle using [Rollup](https://rollupjs.org/)

## 1.0.2

- remove UMD build

## 1.0.1

- define module entry points and UMD bundle

## 1.0.0

Publish `omit-empty-es` as a production bundle using [Parcel](https://parceljs.org/).


src/kind-of.spec.js
import kindOf from './kind-of';

describe('null and undefined', () => {
  it('should work for undefined', () => {
    expect(kindOf(undefined)).toEqual('undefined');
  });

  it('should work for null', () => {
    expect(kindOf(null)).toEqual('null');
  });
});

describe('primitives', () => {
  it('should work for booleans', () => {
    expect(kindOf(true)).toEqual('boolean');
    expect(kindOf(false)).toEqual('boolean');
    expect(kindOf(new Boolean(true))).toEqual('boolean');
  });

  it('should work for numbers', () => {
    expect(kindOf(42)).toEqual('number');
    expect(kindOf(new Number(42))).toEqual('number');
  });

  it('should work for strings', () => {
    expect(kindOf('str')).toEqual('string');
    expect(kindOf(new String('str'))).toEqual('string');
  });
});

describe('objects', () => {
  it('should work for arguments', () => {
    (function iife() {
      // eslint-disable-next-line prefer-rest-params
      expect(kindOf(arguments)).toEqual('arguments');
    })();
  });

  it('should work for buffers', () => {
    expect(kindOf(Buffer.from(''))).toEqual('buffer');
  });

  it('should work for objects', () => {
    function Test() {}
    const instance = new Test();
    const literal = {};
    const createdNull = Object.create(null);
    const createdObj = Object.create({});

    expect(kindOf(instance)).toEqual('object');
    expect(kindOf(literal)).toEqual('object');
    expect(kindOf(createdNull)).toEqual('object');
    expect(kindOf(createdObj)).toEqual('object');
  });

  it('should work for dates', () => {
    expect(kindOf(new Date())).toEqual('date');
  });

  it('should work for arrays', () => {
    /* eslint no-array-constructor: 0 */
    expect(kindOf([])).toEqual('array');
    expect(kindOf([1, 2, 3])).toEqual('array');
    expect(kindOf(new Array())).toEqual('array');
  });

  it('should work for regular expressions', () => {
    expect(kindOf(/./)).toEqual('regexp');
    expect(kindOf(new RegExp('^foo$'))).toEqual('regexp');
  });

  it('should work for functions', () => {
    expect(kindOf(function _() {})).toEqual('function');
    // eslint-disable-next-line no-new-func
    expect(kindOf(new Function())).toEqual('function');
  });

  it('should work for Errors', () => {
    expect(kindOf(new Error(''))).toEqual('error');
  });
});

describe('es6 features', () => {
  it('should work for resolved promises', () => {
    const promise = Promise.resolve(123);
    expect(kindOf(promise)).toEqual('promise');
  });

  it('should work for rejected promises', () => {
    const promise = Promise.reject(new Error('foo bar'));
    promise.catch(function _() {});
    expect(kindOf(promise)).toEqual('promise');
  });

  it('should work for generator functions', () => {
    // eslint-disable-next-line require-yield
    const gen = function* named() {
      return true;
    };
    expect(kindOf(gen)).toEqual('generatorfunction');
  });

  it('should work for generator objects', () => {
    // eslint-disable-next-line require-yield
    const gen = function* named() {
      return true;
    };
    expect(kindOf(gen())).toEqual('generator');
  });

  it('should work for template strings', () => {
    /* eslint quotes: 0 */
    const name = 'Foo';
    expect(kindOf(`Welcome ${name} buddy`)).toEqual('string');
  });

  it('should work for Map', () => {
    const map = new Map();
    expect(kindOf(map)).toEqual('map');
    expect(kindOf(map.set)).toEqual('function');
    expect(kindOf(map.get)).toEqual('function');
    expect(kindOf(map.add)).toEqual('undefined');
  });

  it('should work for WeakMap', () => {
    const weakmap = new WeakMap();
    expect(kindOf(weakmap)).toEqual('weakmap');
    expect(kindOf(weakmap.set)).toEqual('function');
    expect(kindOf(weakmap.get)).toEqual('function');
    expect(kindOf(weakmap.add)).toEqual('undefined');
  });

  it('should work for Set', () => {
    const set = new Set();
    expect(kindOf(set)).toEqual('set');
    expect(kindOf(set.add)).toEqual('function');
    expect(kindOf(set.set)).toEqual('undefined');
    expect(kindOf(set.get)).toEqual('undefined');
  });

  it('should work for WeakSet', () => {
    const weakset = new WeakSet();
    expect(kindOf(weakset)).toEqual('weakset');
    expect(kindOf(weakset.add)).toEqual('function');
    expect(kindOf(weakset.set)).toEqual('undefined');
    expect(kindOf(weakset.get)).toEqual('undefined');
  });

  it('should work for Set Iterator', () => {
    const SetValuesIterator = new Set().values();
    expect(kindOf(SetValuesIterator)).toEqual('setiterator');
  });
  it('should work for Map Iterator', () => {
    const MapValuesIterator = new Map().values();
    expect(kindOf(MapValuesIterator)).toEqual('mapiterator');
  });
  it('should work for Array Iterator', () => {
    const ArrayEntriesIterator = [].entries();
    expect(kindOf(ArrayEntriesIterator)).toEqual('arrayiterator');
  });
  it('should work for String Iterator', () => {
    const StringCharIterator = ''[Symbol.iterator]();
    expect(kindOf(StringCharIterator)).toEqual('stringiterator');
  });

  it('should work for Symbol', () => {
    expect(kindOf(Symbol('foo'))).toEqual('symbol');
    expect(kindOf(Symbol.prototype)).toEqual('symbol');
  });

  it('should work for Int8Array', () => {
    const int8array = new Int8Array();
    expect(kindOf(int8array)).toEqual('int8array');
  });

  it('should work for Uint8Array', () => {
    const uint8array = new Uint8Array();
    expect(kindOf(uint8array)).toEqual('uint8array');
  });

  it('should work for Uint8ClampedArray', () => {
    const uint8clampedarray = new Uint8ClampedArray();
    expect(kindOf(uint8clampedarray)).toEqual('uint8clampedarray');
  });

  it('should work for Int16Array', () => {
    const int16array = new Int16Array();
    expect(kindOf(int16array)).toEqual('int16array');
  });

  it('should work for Uint16Array', () => {
    const uint16array = new Uint16Array();
    expect(kindOf(uint16array)).toEqual('uint16array');
  });

  it('should work for Int32Array', () => {
    const int32array = new Int32Array();
    expect(kindOf(int32array)).toEqual('int32array');
  });

  it('should work for Uint32Array', () => {
    const uint32array = new Uint32Array();
    expect(kindOf(uint32array)).toEqual('uint32array');
  });

  it('should work for Float32Array', () => {
    const float32array = new Float32Array();
    expect(kindOf(float32array)).toEqual('float32array');
  });

  it('should work for Float64Array', () => {
    const float64array = new Float64Array();
    expect(kindOf(float64array)).toEqual('float64array');
  });
});


src/kind-of.spec.ts
import kindOf from './kind-of';

describe('null and undefined', () => {
  it('should work for undefined', () => {
    expect(kindOf(undefined)).toEqual('undefined');
  });

  it('should work for null', () => {
    expect(kindOf(null)).toEqual('null');
  });
});

describe('primitives', () => {
  it('should work for booleans', () => {
    expect(kindOf(true)).toEqual('boolean');
    expect(kindOf(false)).toEqual('boolean');
    expect(kindOf(new Boolean(true))).toEqual('boolean');
  });

  it('should work for numbers', () => {
    expect(kindOf(42)).toEqual('number');
    expect(kindOf(new Number(42))).toEqual('number');
  });

  it('should work for strings', () => {
    expect(kindOf('str')).toEqual('string');
    expect(kindOf(new String('str'))).toEqual('string');
  });
});

describe('objects', () => {
  it('should work for arguments', () => {
    (function iife() {
      // eslint-disable-next-line prefer-rest-params
      expect(kindOf(arguments)).toEqual('arguments');
    })();
  });

  it('should work for buffers', () => {
    expect(kindOf(Buffer.from(''))).toEqual('buffer');
  });

  it('should work for objects', () => {
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    function Test() {}
    // @ts-expect-error
    const instance = new Test();
    const literal = {};
    const createdNull = Object.create(null);
    const createdObj = Object.create({});

    expect(kindOf(instance)).toEqual('object');
    expect(kindOf(literal)).toEqual('object');
    expect(kindOf(createdNull)).toEqual('object');
    expect(kindOf(createdObj)).toEqual('object');
  });

  it('should work for dates', () => {
    expect(kindOf(new Date())).toEqual('date');
  });

  it('should work for arrays', () => {
    /* eslint no-array-constructor: 0 */
    expect(kindOf([])).toEqual('array');
    expect(kindOf([1, 2, 3])).toEqual('array');
    // eslint-disable-next-line @typescript-eslint/no-array-constructor
    expect(kindOf(new Array())).toEqual('array');
  });

  it('should work for regular expressions', () => {
    expect(kindOf(/./)).toEqual('regexp');
    expect(kindOf(new RegExp('^foo$'))).toEqual('regexp');
  });

  it('should work for functions', () => {
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    expect(kindOf(function _() {})).toEqual('function');
    // eslint-disable-next-line no-new-func
    expect(kindOf(new Function())).toEqual('function');
  });

  it('should work for Errors', () => {
    expect(kindOf(new Error(''))).toEqual('error');
  });
});

describe('es6 features', () => {
  it('should work for resolved promises', () => {
    const promise = Promise.resolve(123);
    expect(kindOf(promise)).toEqual('promise');
  });

  it('should work for rejected promises', () => {
    const promise = Promise.reject(new Error('foo bar'));
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    promise.catch(function _() {});
    expect(kindOf(promise)).toEqual('promise');
  });

  it('should work for generator functions', () => {
    // eslint-disable-next-line require-yield
    const gen = function* named() {
      return true;
    };
    expect(kindOf(gen)).toEqual('generatorfunction');
  });

  it('should work for generator objects', () => {
    // eslint-disable-next-line require-yield
    const gen = function* named() {
      return true;
    };
    expect(kindOf(gen())).toEqual('generator');
  });

  it('should work for template strings', () => {
    /* eslint quotes: 0 */
    const name = 'Foo';
    expect(kindOf(`Welcome ${name} buddy`)).toEqual('string');
  });

  it('should work for Map', () => {
    const map = new Map();
    expect(kindOf(map)).toEqual('map');
    expect(kindOf(map.set)).toEqual('function');
    expect(kindOf(map.get)).toEqual('function');
    // @ts-expect-error
    expect(kindOf(map.add)).toEqual('undefined');
  });

  it('should work for WeakMap', () => {
    const weakmap = new WeakMap();
    expect(kindOf(weakmap)).toEqual('weakmap');
    expect(kindOf(weakmap.set)).toEqual('function');
    expect(kindOf(weakmap.get)).toEqual('function');
    // @ts-expect-error
    expect(kindOf(weakmap.add)).toEqual('undefined');
  });

  it('should work for Set', () => {
    const set = new Set();
    expect(kindOf(set)).toEqual('set');
    expect(kindOf(set.add)).toEqual('function');
    // @ts-expect-error
    expect(kindOf(set.set)).toEqual('undefined');
    // @ts-expect-error
    expect(kindOf(set.get)).toEqual('undefined');
  });

  it('should work for WeakSet', () => {
    const weakset = new WeakSet();
    expect(kindOf(weakset)).toEqual('weakset');
    expect(kindOf(weakset.add)).toEqual('function');
    // @ts-expect-error
    expect(kindOf(weakset.set)).toEqual('undefined');
    // @ts-expect-error
    expect(kindOf(weakset.get)).toEqual('undefined');
  });

  it('should work for Set Iterator', () => {
    const SetValuesIterator = new Set().values();
    expect(kindOf(SetValuesIterator)).toEqual('setiterator');
  });
  it('should work for Map Iterator', () => {
    const MapValuesIterator = new Map().values();
    expect(kindOf(MapValuesIterator)).toEqual('mapiterator');
  });
  it('should work for Array Iterator', () => {
    const ArrayEntriesIterator = [].entries();
    expect(kindOf(ArrayEntriesIterator)).toEqual('arrayiterator');
  });
  it('should work for String Iterator', () => {
    const StringCharIterator = ''[Symbol.iterator]();
    expect(kindOf(StringCharIterator)).toEqual('stringiterator');
  });

  it('should work for Symbol', () => {
    expect(kindOf(Symbol('foo'))).toEqual('symbol');
    expect(kindOf(Symbol.prototype)).toEqual('symbol');
  });

  it('should work for Int8Array', () => {
    const int8array = new Int8Array();
    expect(kindOf(int8array)).toEqual('int8array');
  });

  it('should work for Uint8Array', () => {
    const uint8array = new Uint8Array();
    expect(kindOf(uint8array)).toEqual('uint8array');
  });

  it('should work for Uint8ClampedArray', () => {
    const uint8clampedarray = new Uint8ClampedArray();
    expect(kindOf(uint8clampedarray)).toEqual('uint8clampedarray');
  });

  it('should work for Int16Array', () => {
    const int16array = new Int16Array();
    expect(kindOf(int16array)).toEqual('int16array');
  });

  it('should work for Uint16Array', () => {
    const uint16array = new Uint16Array();
    expect(kindOf(uint16array)).toEqual('uint16array');
  });

  it('should work for Int32Array', () => {
    const int32array = new Int32Array();
    expect(kindOf(int32array)).toEqual('int32array');
  });

  it('should work for Uint32Array', () => {
    const uint32array = new Uint32Array();
    expect(kindOf(uint32array)).toEqual('uint32array');
  });

  it('should work for Float32Array', () => {
    const float32array = new Float32Array();
    expect(kindOf(float32array)).toEqual('float32array');
  });

  it('should work for Float64Array', () => {
    const float64array = new Float64Array();
    expect(kindOf(float64array)).toEqual('float64array');
  });
});


src/omit-empty.spec.js
import omitEmpty from './index';

describe('omit-empty', () => {
  it('should return non-empty string values', () => {
    expect(omitEmpty('foo')).toBe('foo');
  });

  it('should return undefined when the value is an empty string', () => {
    expect(omitEmpty('')).toBe(void 0);
  });

  it('should return non-empty arrays', () => {
    expect(omitEmpty(['foo'])).toEqual(['foo']);
  });

  it('should omit empty values from the given object', () => {
    expect(omitEmpty({ one: {}, a: '', b: 'c' })).toEqual({ b: 'c' });
    expect(omitEmpty({ a: undefined, b: 'c' })).toEqual({ b: 'c' });
    expect(omitEmpty({ a: null, b: 'c' })).toEqual({ b: 'c' });
    expect(omitEmpty({ a: '', b: 'c' })).toEqual({ b: 'c' });
  });

  it('should omit deeply nested empty values from the given object', () => {
    const fixture = {
      foo: [{ a: '' }, { bar: 'baz' }, [{ a: '' }, { bar: 'baz' }]],
      one: {
        two: { three: { four: { abc: { xyz: '' } } }, five: '', six: 'seven' },
      },
      a: '',
      b: 'c',
    };

    expect(omitEmpty(fixture)).toEqual({
      foo: [{ bar: 'baz' }, [{ bar: 'baz' }]],
      one: { two: { six: 'seven' } },
      b: 'c',
    });
  });

  it('should omit empty objects.', () => {
    expect(omitEmpty({ a: { b: { c: 'foo' }, d: {} } })).toEqual({
      a: { b: { c: 'foo' } },
    });
  });

  it('should omit nested empty objects.', () => {
    expect(omitEmpty({ a: { b: undefined, c: 'd' } })).toEqual({
      a: { c: 'd' },
    });
    expect(omitEmpty({ a: { b: null, c: 'd' } })).toEqual({ a: { c: 'd' } });
    expect(omitEmpty({ a: { b: '', c: 'd' } })).toEqual({ a: { c: 'd' } });
  });

  it('should deeply omit nested empty objects.', () => {
    expect(omitEmpty({ a: { b: undefined, c: void 0 } })).toEqual({});
    expect(omitEmpty({ a: { b: null, c: void 0 } })).toEqual({});
    expect(omitEmpty({ a: { b: '', c: void 0 } })).toEqual({});
  });

  it('should not omit functions', () => {
    // eslint-disable-next-line no-unused-vars
    const fn = (a, b, c) => {};
    const fn2 = () => {};
    expect(omitEmpty({ a: fn, b: fn2 })).toEqual({ a: fn, b: fn2 });
  });

  it('should omit empty strings.', () => {
    expect(omitEmpty({ a: { b: { c: 'foo' }, d: '' } })).toEqual({
      a: { b: { c: 'foo' } },
    });
  });

  it('should omit empty Map', () => {
    expect(omitEmpty({ a: { b: { c: 'foo' }, d: new Map() } })).toEqual({
      a: { b: { c: 'foo' } },
    });
  });

  it('should omit empty Set', () => {
    expect(omitEmpty({ a: { b: { c: 'foo' }, d: new Set() } })).toEqual({
      a: { b: { c: 'foo' } },
    });
  });

  it('should not omit regex', () => {
    expect(omitEmpty({ a: { b: { c: /foo/ }, d: new Set() } })).toEqual({
      a: { b: { c: /foo/ } },
    });
  });

  it('should omit empty arrays.', () => {
    expect(omitEmpty({ a: { b: { c: 'foo', d: [] }, foo: [] } })).toEqual({
      a: { b: { c: 'foo' } },
    });
    expect(
      omitEmpty({ a: { b: { c: 'foo', d: [void 0] }, foo: [null] } })
    ).toEqual({
      a: { b: { c: 'foo' } },
    });
    expect(omitEmpty({ a: { b: { c: 'foo', d: [''] }, foo: [null] } })).toEqual(
      {
        a: { b: { c: 'foo' } },
      }
    );
    expect(
      omitEmpty({ a: { z: [''], b: { c: 'foo', d: [''] }, foo: [null] } })
    ).toEqual({ a: { b: { c: 'foo' } } });
  });

  it('should not omit zero', () => {
    const actual = omitEmpty({ a: { b: { c: 'foo', d: 0 }, foo: [] } });
    const expected = { a: { b: { c: 'foo', d: 0 } } };
    expect(actual).toEqual(expected);
  });

  it('should omit zero when omitZero is true', () => {
    const expected = { a: { b: { c: 'foo' } } };
    const actual = omitEmpty(
      { a: { b: { c: 'foo', d: 0 }, foo: [] } },
      { omitZero: true }
    );
    expect(actual).toEqual(expected);
  });

  it('should not omit boolean false', () => {
    const actual = omitEmpty({
      a: { b: { c: 'foo', d: 0 }, foo: [], bar: false },
    });
    const expected = { a: { b: { c: 'foo', d: 0 }, bar: false } };
    expect(actual).toEqual(expected);
  });

  it('should not omit Dates', () => {
    const today = new Date();
    const actual = omitEmpty({
      a: { b: { c: 'foo', d: today }, foo: [], bar: false },
    });
    const expected = { a: { b: { c: 'foo', d: today }, bar: false } };
    expect(actual).toEqual(expected);
  });

  it('should omit deeply nested values', () => {
    const o = {
      a: {
        b: { c: 'foo', d: 0, e: { f: { g: {}, h: { i: 'i' } } } },
        foo: [['bar', 'baz'], []],
        bar: [],
        one: 1,
        two: 2,
        three: 0,
      },
    };

    expect(omitEmpty(o)).toEqual({
      a: {
        b: { c: 'foo', d: 0, e: { f: { h: { i: 'i' } } } },
        foo: [['bar', 'baz']],
        one: 1,
        two: 2,
        three: 0,
      },
    });
  });
});


.eslintrc.js
/**
 * @type {import("eslint").Linter.Config}
 */
module.exports = {
  parser: '@typescript-eslint/parser',
  extends: [
    'plugin:@typescript-eslint/recommended',
    'plugin:import/errors',
    'plugin:import/warnings',
    'plugin:import/typescript',
    'plugin:jest/recommended',
    'prettier',
  ],
  plugins: ['import', 'jest', 'prettier', '@typescript-eslint/eslint-plugin'],
  rules: {
    '@typescript-eslint/ban-types': 0,
    '@typescript-eslint/naming-convention': 0,
    '@typescript-eslint/consistent-type-definitions': 0,
    '@typescript-eslint/no-explicit-any': 2,
    '@typescript-eslint/no-use-before-define': [
      'error',
      {
        functions: false,
      },
    ],
    '@typescript-eslint/no-var-requires': 0,
    '@typescript-eslint/unbound-method': 0,
    'import/no-extraneous-dependencies': 0,
    'import/no-named-as-default': 0,
    'import/no-named-as-default-member': 0,
    'import/default': 0,
    'import/no-unresolved': 2,
  },
  settings: {
    'import/parsers': {
      '@typescript-eslint/parser': ['.js', '.jsx', '.ts', '.tsx'],
    },
    'import/resolver': {
      'eslint-import-resolver-typescript': true,
      typescript: {},
      node: {
        extensions: ['.js', '.jsx', '.ts', '.tsx'],
      },
    },
  },
  overrides: [
    {
      files: ['*.spec.ts'],
      env: {
        jest: true,
      },
    },
    {
      files: ['*.ts', '*.tsx'],
      rules: {
        '@typescript-eslint/ban-ts-comment': 0,
        '@typescript-eslint/explicit-function-return-type': 0,
        '@typescript-eslint/explicit-member-accessibility': [
          2,
          {
            accessibility: 'no-public',
          },
        ],
        '@typescript-eslint/no-require-imports': 0,
        '@typescript-eslint/promise-function-async': 0,
      },
    },
  ],
};


src/kind-of.js
const { toString } = Object.prototype;

function kindOf(val) {
  if (val === void 0) return 'undefined';
  if (val === null) return 'null';

  let type = typeof val;
  if (type === 'boolean') return 'boolean';
  if (type === 'string') return 'string';
  if (type === 'number') return 'number';
  if (type === 'symbol') return 'symbol';
  if (type === 'function') {
    return isGeneratorFn(val) ? 'generatorfunction' : 'function';
  }

  if (isArray(val)) return 'array';
  if (isBuffer(val)) return 'buffer';
  if (isArguments(val)) return 'arguments';
  if (isDate(val)) return 'date';
  if (isError(val)) return 'error';
  if (isRegexp(val)) return 'regexp';

  switch (ctorName(val)) {
    case 'Symbol':
      return 'symbol';
    case 'Promise':
      return 'promise';

    // Set, Map, WeakSet, WeakMap
    case 'WeakMap':
      return 'weakmap';
    case 'WeakSet':
      return 'weakset';
    case 'Map':
      return 'map';
    case 'Set':
      return 'set';

    // 8-bit typed arrays
    case 'Int8Array':
      return 'int8array';
    case 'Uint8Array':
      return 'uint8array';
    case 'Uint8ClampedArray':
      return 'uint8clampedarray';

    // 16-bit typed arrays
    case 'Int16Array':
      return 'int16array';
    case 'Uint16Array':
      return 'uint16array';

    // 32-bit typed arrays
    case 'Int32Array':
      return 'int32array';
    case 'Uint32Array':
      return 'uint32array';
    case 'Float32Array':
      return 'float32array';
    case 'Float64Array':
      return 'float64array';
    default:
      break;
  }

  if (isGeneratorObj(val)) {
    return 'generator';
  }

  // Non-plain objects
  type = toString.call(val);
  switch (type) {
    case '[object Object]':
      return 'object';
    // iterators
    case '[object Map Iterator]':
      return 'mapiterator';
    case '[object Set Iterator]':
      return 'setiterator';
    case '[object String Iterator]':
      return 'stringiterator';
    case '[object Array Iterator]':
      return 'arrayiterator';
    default:
      break;
  }

  // other
  return type
    .slice(8, -1)
    .toLowerCase()
    .replace(/\s/g, '');
}

function ctorName(val) {
  return val.constructor ? val.constructor.name : null;
}

function isArray(val) {
  if (Array.isArray) return Array.isArray(val);
  return val instanceof Array;
}

function isError(val) {
  return (
    val instanceof Error ||
    (typeof val.message === 'string' &&
      val.constructor &&
      typeof val.constructor.stackTraceLimit === 'number')
  );
}

function isDate(val) {
  if (val instanceof Date) return true;
  return (
    typeof val.toDateString === 'function' &&
    typeof val.getDate === 'function' &&
    typeof val.setDate === 'function'
  );
}

function isRegexp(val) {
  if (val instanceof RegExp) return true;
  return (
    typeof val.flags === 'string' &&
    typeof val.ignoreCase === 'boolean' &&
    typeof val.multiline === 'boolean' &&
    typeof val.global === 'boolean'
  );
}

function isGeneratorFn(name) {
  return ctorName(name) === 'GeneratorFunction';
}

function isGeneratorObj(val) {
  return (
    typeof val.throw === 'function' &&
    typeof val.return === 'function' &&
    typeof val.next === 'function'
  );
}

function isArguments(val) {
  try {
    if (typeof val.length === 'number' && typeof val.callee === 'function') {
      return true;
    }
  } catch (err) {
    if (err.message.indexOf('callee') !== -1) {
      return true;
    }
  }
  return false;
}

/**
 * If you need to support Safari 5-7 (8-10 yr-old browser),
 * take a look at https://github.com/feross/is-buffer
 */

function isBuffer(val) {
  if (val.constructor && typeof val.constructor.isBuffer === 'function') {
    return val.constructor.isBuffer(val);
  }
  return false;
}

export default kindOf;


babel.config.js
/**
 * @type {import('@babel/core').ConfigFunction}
 */
module.exports = function getBabelPreset(api) {
  if (api) {
    // Cache the returned value forever and don't call this function again.
    api.cache(true);
  }

  // This is similar to how `env` works in Babel:
  // https://babeljs.io/docs/usage/babelrc/#env-option
  // We are not using `env` because it’s ignored in versions > babel-core@6.10.4:
  // https://github.com/babel/babel/issues/4539
  // https://github.com/facebook/create-react-app/issues/720
  // It’s also nice that we can enforce `NODE_ENV` being specified.
  const env = process.env.BABEL_ENV || process.env.NODE_ENV;
  const isEnvDevelopment = env === 'development';
  const isEnvProduction = env === 'production';
  const isEnvTest = env === 'test';

  if (!isEnvDevelopment && !isEnvProduction && !isEnvTest) {
    throw new Error(
      'The babel preset of requires that you specify `NODE_ENV` or ' +
        '`BABEL_ENV` environment variables. Valid values are "development", ' +
        `"test", and "production". Instead, received: ${JSON.stringify(env)}.`
    );
  }

  /**
   * @type {import('@babel/core').TransformOptions}
   */
  return {
    presets: [
      isEnvTest && [
        // ES features necessary for user's Node version
        require('@babel/preset-env').default,
        {
          targets: {
            browsers: ['last 2 versions'],
            node: 'current',
          },
        },
      ],
      (isEnvProduction || isEnvDevelopment) && [
        // Latest stable ECMAScript features
        require('@babel/preset-env').default,
        {
          targets: {
            browsers: ['last 2 versions'],
          },
          corejs: { version: 3, proposals: true },
          // `entry` transforms `@babel/polyfill` into individual requires for
          // the targeted browsers. This is safer than `usage` which performs
          // static code analysis to determine what's required.
          // This is probably a fine default to help trim down bundles when
          // end-users inevitably import '@babel/polyfill'.
          useBuiltIns: 'entry',
          // Do not transform modules to CJS
          modules: false,
          include: ['transform-classes'],
        },
      ],
      [require('@babel/preset-typescript').default],
    ].filter(Boolean),
    plugins: [
      // Polyfills the runtime needed for async/await and generators
      [
        require('@babel/plugin-transform-runtime').default,
        {
          corejs: 3,
          // To be able to use `runtime` in Rollup babel plugin
          helpers: true,
          regenerator: true,
        },
      ],
    ],
  };
};


src/kind-of.ts
const { toString } = Object.prototype;

function kindOf(val: unknown): string {
  if (val === void 0) return 'undefined';
  if (val === null) return 'null';

  let type = typeof val;
  if (type === 'boolean') return 'boolean';
  if (type === 'string') return 'string';
  if (type === 'number') return 'number';
  if (type === 'symbol') return 'symbol';
  if (type === 'function') {
    return isGeneratorFn(val) ? 'generatorfunction' : 'function';
  }

  if (isArray(val)) return 'array';
  if (isBuffer(val)) return 'buffer';
  if (isArguments(val)) return 'arguments';
  if (isDate(val)) return 'date';
  if (isError(val)) return 'error';
  if (isRegexp(val)) return 'regexp';

  switch (ctorName(val)) {
    case 'Symbol':
      return 'symbol';
    case 'Promise':
      return 'promise';

    // Set, Map, WeakSet, WeakMap
    case 'WeakMap':
      return 'weakmap';
    case 'WeakSet':
      return 'weakset';
    case 'Map':
      return 'map';
    case 'Set':
      return 'set';

    // 8-bit typed arrays
    case 'Int8Array':
      return 'int8array';
    case 'Uint8Array':
      return 'uint8array';
    case 'Uint8ClampedArray':
      return 'uint8clampedarray';

    // 16-bit typed arrays
    case 'Int16Array':
      return 'int16array';
    case 'Uint16Array':
      return 'uint16array';

    // 32-bit typed arrays
    case 'Int32Array':
      return 'int32array';
    case 'Uint32Array':
      return 'uint32array';
    case 'Float32Array':
      return 'float32array';
    case 'Float64Array':
      return 'float64array';
    default:
      break;
  }

  if (isGeneratorObj(val)) {
    return 'generator';
  }

  // Non-plain objects
  // @ts-expect-error
  type = toString.call(val);
  switch (type) {
    // @ts-expect-error
    case '[object Object]':
      return 'object';
    // iterators
    // @ts-expect-error
    case '[object Map Iterator]':
      return 'mapiterator';
    // @ts-expect-error
    case '[object Set Iterator]':
      return 'setiterator';
    // @ts-expect-error
    case '[object String Iterator]':
      return 'stringiterator';
    // @ts-expect-error
    case '[object Array Iterator]':
      return 'arrayiterator';
    default:
      break;
  }

  // other
  return type.slice(8, -1).toLowerCase().replace(/\s/g, '');
}

// @ts-expect-error
function ctorName(val) {
  return val.constructor ? val.constructor.name : null;
}

// @ts-expect-error
function isArray(val) {
  if (Array.isArray) return Array.isArray(val);
  return val instanceof Array;
}

// @ts-expect-error
function isError(val) {
  return (
    val instanceof Error ||
    (typeof val.message === 'string' &&
      val.constructor &&
      typeof val.constructor.stackTraceLimit === 'number')
  );
}

// @ts-expect-error
function isDate(val) {
  if (val instanceof Date) return true;
  return (
    typeof val.toDateString === 'function' &&
    typeof val.getDate === 'function' &&
    typeof val.setDate === 'function'
  );
}

// @ts-expect-error
function isRegexp(val) {
  if (val instanceof RegExp) return true;
  return (
    typeof val.flags === 'string' &&
    typeof val.ignoreCase === 'boolean' &&
    typeof val.multiline === 'boolean' &&
    typeof val.global === 'boolean'
  );
}

// @ts-expect-error
function isGeneratorFn(name) {
  return ctorName(name) === 'GeneratorFunction';
}

// @ts-expect-error
function isGeneratorObj(val) {
  return (
    typeof val.throw === 'function' &&
    typeof val.return === 'function' &&
    typeof val.next === 'function'
  );
}

// @ts-expect-error
function isArguments(val) {
  try {
    if (typeof val.length === 'number' && typeof val.callee === 'function') {
      return true;
    }
  } catch (err) {
    if (err instanceof Error && err.message.indexOf('callee') !== -1) {
      return true;
    }
  }
  return false;
}

/**
 * If you need to support Safari 5-7 (8-10 yr-old browser),
 * take a look at https://github.com/feross/is-buffer
 */
// @ts-expect-error
function isBuffer(val) {
  if (val.constructor && typeof val.constructor.isBuffer === 'function') {
    return val.constructor.isBuffer(val);
  }
  return false;
}

export default kindOf;
